teos/src/dbm.rs

```diff
@@ -330,23 +330,32 @@ impl DBM {
     /// matching this locator. If no locator is given, all the appointments in the database would be returned.
     pub(crate) fn load_appointments(
         &self,
-        locator: Option<Locator>,
+        locator_and_userid: Option<(Locator, Option<UserId>)>,
     ) -> HashMap<UUID, ExtendedAppointment> {
         let mut appointments = HashMap::new();
 
         let mut sql =
             "SELECT a.UUID, a.locator, a.encrypted_blob, a.to_self_delay, a.user_signature, a.start_block, a.user_id
                 FROM appointments as a LEFT JOIN trackers as t ON a.UUID=t.UUID WHERE t.UUID IS NULL".to_string();
+
         // If a locator was passed, filter based on it.
-        if locator.is_some() {
-            sql.push_str(" AND a.locator=(?)");
+        if locator_and_userid.is_some() {
+            sql.push_str(" AND a.locator=(?1)");
+        }
+
+        // If a user_id is passed, filter even more.
+        if locator_and_userid.is_some_and(|inner| inner.1.is_some()) {
+            sql.push_str(" AND a.user_id=(?2)");
         }
```

Comment:
I think this statement  `let mut stmt = self.connection.prepare(&sql).unwrap();` would prevent the addition of `query_args.push(locator.to_vec());` .  I believe only after the initialisation of the `stmt` variable can be query args be added , which  also takes arrays as its parameters. Also the intial changes was made this way, in which i mean joined 

`
if let Some((_, user_id)) = locator_and_userid {
            sql.push_str(" AND a.locator=(?1)");
            if user_id.is_some() {
                sql.push_str(" AND a.user_id=(?2)");
            }
        };
`,  in which Sergio recommended i split to make it cleaner.

Reply:
Yup, this actually wouldn't work since `stmt.query(...)` won't accept a rust vector. I found [this extension](https://docs.rs/rusqlite/0.29.0/rusqlite/vtab/array/index.html) which allows preparing a statement with vector as args, but it's not worth the hustle. I only suggested this to not use the match below but it's not bad anyway.

Regarding nested checks, splitting the checks up in different lines makes sense when the method's args are `Option<Locator>` *and* `Option<UserId>`, because then we would add the appropriate filters without thinking about the dependency between them (we shouldn't nest the checks in this case).

But we went with the `Option<(Locator, Option<UserId>)>` route, in which there is a dependency between the locator and the user id (locator has to exist for user id to exist), so it makes more sense have the check for user id nested inside.

Reply:
Guess I found the review [comment](https://github.com/talaia-labs/rust-teos/pull/237#discussion_r1532399009) you are referring to. Well then, let's keep it as it is now until we all agree on how this should look.

@sr-gi What do you think about the comment above.